### ISSUE TITLE
type-checking input to mxEval

### ISSUE THREAD

**Issue by tbates** [OpenMx/OpenMx#40]
If the user puts the model first, and expression second, they get an error which suggests some component of the expression was not found in the model.

    mxEval(m1, solve(sqrt(I * Vtot)) + ) Error in solve(sqrt(I * Vtot)) : object 'Vtot' not found

Be nice to `is(expression, "MxModel")` and stop if input 1 is an mxModel, or directly test if input1 is a valid expression.

    if(is(expression, "MxModel")){
      stop("Input 1 should be an expression, you gave me a ", class(expression)
    }

A model is not a valid expression:

    mxEval(m1$top, m1)
    Cannot evaluate the object 'top' in the model 'ACE'

**Comment by jpritikin** [comment]
At the beginning of `EvalInternal`, I tried,
````
if (!is(model, "MxModel")) stop("Must be a model")
````
However, this doesn't work because R can't check the type of the `model` argument because it doesn't resolve to a valid symbol,

````
> mxEval(modelOut, test12)
Error in is(model, "MxModel") : object 'test12' not found
````
What else to try?

**Comment by mcneale** [comment]
I would have thought
class(model)
would evaluate to “MxModel” universally, but actually it could be “MxRAMModel” or I suspect a few others.  Still, a test should be easy enough to construct, even if it is just that the first two letters should be Mx and the last 5 Model.

Cheers
Mike

NB More rapid response may be had from mcneale@mac.com<mailto:mcneale@mac.com> - my VCU email is not available on my iPhone or iPad

Michael C. Neale, Ph.D.
Professor, Virginia Institute for Psychiatric and Behavioral Genetics
Virginia Commonwealth University
Shipping Address: 800 East Leigh St. Suite 1-156, Richmond VA 23219-1534
Mailing Address : Box 980126 Richmond VA 23298-0126
Telephone: (804) 828-3369  mailto:mcneale@mac.com

On Oct 30, 2017, at 12:34 PM, Joshua Pritikin <notifications@github.com<mailto:notifications@github.com>> wrote:


At the beginning of EvalInternal, I tried,

if (!is(model, "MxModel")) stop("Must be a model")


However, this doesn't work because R can't check the type of the model argument because it doesn't resolve to a valid symbol,

> mxEval(modelOut, test12)
Error in is(model, "MxModel") : object 'test12' not found


What else to try?

**Comment by tbates** [comment]
Solution might be to test position1 not position2: i.e.,  expression = , not  model = . 

Even though what we might wish the user to do is put a model in place 2, If place1 (expression = ) is a model, we can stop.

And that should work, as we avoid the problem of the user's desired expression (offered up as a model) not being  resolvable.

    if (!is(expression, "MxModel")) stop("Expression should come first - you probably put the  model first")

**Comment by mcneale** [comment]
Yeah - but how often do users make this kind of mistake?

**Comment by tbates** [comment]
tldr: input checking saves users a lot of confusion and time, and costs us not much.

Without input checking, users get messages that are impossible to decipher, because the wrong object was offered up. Especially for a package like OpenMx where users find themselves coping with far more warnings and errors than they're used to in other packages, and it's trivial to check the input type, I think we should do it.

In this case, widely used packages differ in whether functions which involve data and some action go in the order fn(data, question) or fn(question, data), so I'm guessing this happens often enough.

**Comment by mcneale** [comment]
Agreed, always good to spot errors and give interpretable feedback.  R as a whole is crap at this, but that doesn’t mean OpenMx has to be.  Making a human interface to R errors seems like a worthwhile project, really.

**Comment by tbates** [comment]
Shared sentiment -- everything we don't capture gets sent to R, which says "ha ha! figure this out " :-)

**Comment by RMKirkpatrick** [comment]
> Without input checking, users get messages that are impossible to decipher, because the wrong object was offered up. Especially for a package like OpenMx where users find themselves coping with far more warnings and errors than they're used to in other packages, and it's trivial to check the input type, I think we should do it.

I completely agree, and I don't think OpenMx's functions do input-checking as often as they should.

**Comment by tbates** [comment]
The solution to this seems to be in the file `MxEval.R` to put the check into the `mxEval` function, not `EvalInternal` 

Something like:

```R
mxEval <- function(expression, model, compute = FALSE, show = FALSE, defvar.row = 1L,
			cache = new.env(parent = emptyenv()), cacheBack = FALSE) {
	if (missing(expression)) {
		stop("'expression' argument is mandatory in call to mxEval function")
	} else if (missing(model)) {
		stop("'model' argument is mandatory in call to mxEval function")
	}
	 if(!is(expression, "MxModel")){
		stop("You might have put the model in the first instead of the second parameter of mxEval")
	}
	expression <- match.call()$expression
	modelvariable <- match.call()$model
	EvalInternal(expression, model, modelvariable, compute, show, defvar.row,
		       cache, cacheBack, 2L)
}

```

**Comment by jpritikin** [comment]
> The solution to this seems to be in the file MxEval.R to put the check into the mxEval function, not EvalInternal

I think you added `!` by mistake. Here's the proposed change:

```
diff --git a/R/MxEval.R b/R/MxEval.R
index 17165930c..72bbd279e 100755
--- a/R/MxEval.R
+++ b/R/MxEval.R
@@ -20,6 +20,9 @@ mxEval <- function(expression, model, compute = FALSE, show = FALSE, defvar.row
        } else if (missing(model)) {
                stop("'model' argument is mandatory in call to mxEval function")
        }
+       if(is(expression, "MxModel")){
+               stop("You might have put the model in the first instead of the second parameter of mxEval")
+       }
        expression <- match.call()$expression
        modelvariable <- match.call()$model
        EvalInternal(expression, model, modelvariable, compute, show, defvar.row,
```

Among many failures that this causes, `inst/models/passing/defvaralgebra.R` fails with,

```
Error in B[1, 1] : object of type 'S4' is not subsettable
Calls: omxCheckEquals ... checkEqualDimensions -> checkZeroDimensions -> mxEval -> is
```

If would save time if, prior to proposing any solution, the advocate would check whether the proposal passes `make test`.

**Comment by tbates** [comment]
Apologies, I didn't intend to commit this... A mistake on my part: intended
to simply ask about it here.

**Comment by tbates** [comment]
I don’t see having added this -t’s uncommitted on my end..

**Comment by jpritikin** [comment]
> I don’t see having added this -t’s uncommitted on my end..

Yeah, that was my point. It was such a simple change that you should have tested it yourself, but you didn't.

**Comment by tbates** [comment]
Just as an example of the value of input checking: Question from a student stymied for a few days about what was here:

```R
mxSE(c3a, "a_std")
```
> Error in mxSE(c3a, "a_std") : 
  trying to get slot "output" from an object of a basic class ("character") with no slots

**Comment by mcneale** [comment]
Agreed, it seems that we need a standardized form of input checking - a function to find out what kind of object it is, what it is supposed to be, and that reports the error in a friendly way.

"Terribly sorry!  I can't apply this function to the arguments you supplied because at least one of them is of the wrong type.  Specifically, the argument <argname> was expecting an object of class <whatever> but instead you provided it with an object of class <whatever it actually was>.  Different strokes for different folks and all..."

If we were being extra nice, we might check all the arguments and list all the ones that the user SNAFU'd.
